### ISSUE TITLE
Study sky background in Twinkles outputs and compare them to backgrounds in Opsim/SkyBrightness Model etc.

### ISSUE THREAD

**Issue by rbiswas4** [DarkEnergyScienceCollaboration/Twinkles#205]
- [] Obtain the sky brightness background from visits by 
    - [] obtaining the background image for each visit
    - [] obtain the median values for sky background
    - [] compare to OpSim filtskybrightness for the same visit (and similar predicted quantities)

@jbkalmbach  @jchiang87 
However, I am a little confused right now as to what should do what. 

Should we be trying to obtain all these quantities and add them to the database that @jchiang87  has started work on? I think this would make most sense, but I may be missing something. Is there good reason not to put visit specific information in this database or is it already there?

**Comment by jchiang87** [comment]
@rbiswas4 That information should go into the CcdVisit table, which is created by the `db_table_access.py` code using this schema:  

https://lsst-web.ncsa.illinois.edu/schema/index.php?sVer=baseline&t=CcdVisit

You could either update the [CcdVisitTable](https://github.com/DarkEnergyScienceCollaboration/Twinkles/blob/master/python/desc/twinkles/db_table_access.py#L110) class, adding a method to ingest that info, or write a separate script to do the inserts.   I guess the former would be the most sensible thing.

Right now, the `CcdVisitTable` code is getting the basic visit info from the `registry.sqlite3` file, but it could get it from some other source.  I tried to understand what the [ingestProcessed.py](https://github.com/lsst/datarel/blob/master/bin/ingest/ingestProcessed.py) script from the `datarel` package was meant to do in this regard, but it looked too complicated for my immediate purposes.

**Comment by drphilmarshall** [comment]
Are you saying that the Twinkles `CcdVisitTable` does not currently contain a `skyBg` column, because we did not run the `ingestProcessed.py` script? I see from the schema that the `CcdVisitTable` is supposed to contain these three interesting quantities, among other things: 
```
seeing	    FLOAT	y	arcsec		Mean measured FWHM of the PSF.
skyBg	    FLOAT	y	DN		    Average sky background.
skyNoise	FLOAT	y	DN		    RMS noise of the sky background.
```
Are the `seeing` and `skyNoise` also currently missing?

**Comment by jchiang87** [comment]
It's got those columns, but they're filled with NULL currently.   The only ones that are being filled with meaningful values (from the registry.sqlite3 file) are ccdVisitId, visitId, ccdName, raftName, filterName, and obsStart.  I don't think the `ingestProcessed.py` script will work out of the box, or at least not without some head-scratching.

**Comment by drphilmarshall** [comment]
Got it, thanks! Rahul, do you have a JimDB already made, up there? I guess you'll need one to work on.

**Comment by rbiswas4** [comment]
@drphilmarshall @jchiang87 No, I do not have a JimDB made up yet. I want to first make sure I am doing the right stuff and then ingest this into JimDB. @jchiang87's pointing out the schema is quite useful.

**Comment by rbiswas4** [comment]
So, from a conceptual point of view, we have the following classes of information:

- Forced Sources:  For each visit, this has measurements for each object identified as a Source. eg. Flux, Flux error
- Visit Related Information : eg. seeing, PSF width, zero point etc, sky background

Is this why there are two tables? So, the reason for ccdvisit ID is to create an integer ID using the dictID?

**Comment by jchiang87** [comment]
Yes, that's correct.  Although, there is a [Visit](https://lsst-web.ncsa.illinois.edu/schema/index.php?sVer=baseline&t=Visit) table, which has the filterName, obsStart, expMidPt, expTime columns.  If the database tables were completely normalized, those columns wouldn't also appear in CcdVisit.  And since I wanted to plot source locations, I implemented the Object table based on the merged catalog from the coadded detections.

**Comment by rbiswas4** [comment]
@jchiang87 
The information that I think we need to be able to pull out of (joins) of the databases:

objectid ccdvisitID visitID mjd ra dec ra_sigma dec_sigma pixel_x pixel_y flux flux_err zero point zero point Err skyBackground

objectID : ID assigned to detected objects on which forced photometry is being run
visitID : ID that indexes every single visit (OpSim obsHistID)
ccdVisitID : ID that indexes a particular location of the telescope as well as visitID (right now this is equivalent to visitID for us, because we have a single CCD)
ra : ra for the object
dec : dec for the object
ra_sigma :  uncertainties on ra 
dec_sigma : uncertainties on dec
flux : type of flux, unit
flux_sigma : type of flux, unit
bandpass : filter
zero_point : zero point, (probably to define the unit)
zero_point error : uncertainty of zero point
Sky Background : value 

I am able to get most of these quantities through the butler. So, I think I would like to finally update your ingestion scripts. 

However I am somewhat confused about the following : 

The forcedsources directory has a lot of other quantities (many of which are different definitions of flux). Are we planning  to get rid of these quantities in the database? Or should we just collate everything into a database and decide what we want later? I think the latter is easiler (and I have something that would do this). So I was wondering if you would prefer not to: (maybe space issues)?

**Comment by jchiang87** [comment]
I think it would be fine to add those other quantities.  At the time, I was just interested in implementing the  schema for the ForcedSource table as shown at that [LSST Schema Browser](https://lsst-web.ncsa.illinois.edu/schema/index.php?sVer=baseline&t=ForcedSource) site.  It only has 5 columns, so I was a bit relieved not have to transcribe a lot of column in the `create table` query.   But since we don't really know what will be useful from those catalogs, we might as well add everything that looks reasonable.

**Comment by jchiang87** [comment]
Hi @rbiswas4 ,

I implemented some straw-man calculations of `seeing`, `skyBg`, and `skyNoise` using the `calexp` products and update those values in the `CcdVisit` table.  The code is in `pserv` [here](https://github.com/DarkEnergyScienceCollaboration/pserv/blob/master/python/desc/pserv/utils.py#L51).   Could you have a look and comment?   Thanks!

**Comment by rbiswas4** [comment]
Hi @jchiang87 

I am traveling and will miss tomorrow's meeting. I looked at the code and it seems good. The one thing that seems like a typo is that a square root seems necessary for sky noise. 

One thing that I think we may want to do is find these values of skybg in units of mag/arcsec^2 rather than counts. I understand your choice follows the schema of ccdvisit table, but this would be a useful quantity for us. I can see that we have three choices:
- Change the schema units
- add this as a function so that it is easily available
- add a new column

I think 2 might be the best option if that can be set up, but am quite open to other options. 

I had code here : 
https://github.com/DarkEnergyScienceCollaboration/pserv/blob/master/python/desc/pserv/utils.py#L51
and here to do this
https://github.com/DarkEnergyScienceCollaboration/pserv/blob/smalldb/python/desc/pserv/ingestRepo.py#L64
that pertains to this.

Finally, I am also picking up pep8 violations and unused imports of afwImage and DBConnection. 


FYI : this code is in a branch I was setting up to see if I could set up automated ingests of quantities mostly to explore how we should construct databases with quantities on sim side which are not in the schema ... but that does not seem to be going in the direction I had hoped, so I do not plan to try to merge this branch and just pushed it github so that the relevant code can be pulled out if desired.

**Comment by jchiang87** [comment]
@rbiswas4 Thanks for having a look at the code!  I agree that adding a function to compute the skybg in terms of mag/arcsec^2 is the best way to go for now.  I'll also make the fixes you noted (and will fill the zeroPoint column.)

> FYI : this code is in a branch I was setting up to see if I could set up automated ingests of quantities mostly to explore how we should construct databases with quantities on sim side which are not in the schema ... but that does not seem to be going in the direction I had hoped, so I do not plan to try to merge this branch and just pushed it github so that the relevant code can be pulled out if desired.

Are there specific features you were considering?  From conversations with the Qserv team, they were recommending against modifying their schema and suggested instead to create separate tables with the quantities we want and do joins with their tables.  How this would work once we have a Qserv db to use is unclear, but since Pserv is just MySQL, this would be straight-forward to implement.

**Comment by rbiswas4** [comment]
> Are there specific features you were considering? From conversations with the Qserv team, they were recommending against modifying their schema and suggested instead to create separate tables with the quantities we want and do joins with their tables. How this would work once we have a Qserv db to use is unclear, but since Pserv is just MySQL, this would be straight-forward to implement.

I was under the impression that the baseline schema was very fluid, so good to know. Regardless, it would simply make sense to keep these quantities in separate tables, which is alsp what I ad in mind. The feature that I am interested in is making this easy rather than having to write out a schema like you did, thereby making the addition of new quantities into tables very easy. Of course, that does require some work elsewhere: and if you noticed the functions I pointed you to, it had more information than the quantity itself (quantity, name , type)  so that we could easily add to the schema set of these quantities when needed. 

So, I would like the functionality that a user can use a function to set up a database that includes 
columns like 
obsHistID, skyBG, 
and another can easily use the same code to set up
obsHistID, zeropt 
if necessary, without having to write out the schema for his or her chosen set. 

I am not sure how well this would work, or if there is a better way of achieving this and it would be great to hear if you have opinions on this.

**Comment by jchiang87** [comment]
For those examples, I can imagine a table with obsHistID as the primary key, and having tools to add columns as desired, e.g., the table may initially just have obsHistID and skyBG as columns, then later someone could run the code to add the zeropt column and fill it.   I played around with this before and [this recipe](http://stackoverflow.com/questions/5677932/optimize-mysql-for-faster-alter-table-add-column/5678073#5678073) seemed like an efficient way to implement that.    We might want to think about which columns would be appropriate primary keys.

I think this would be a useful a facility since people will undoubtedly want tables with their own computed values that they could join with the standard ones and having a programmatic interface like this might help manage the proliferation of new tables.   Maybe we could write some test code that exercises this functionality so we know what to implement.

**Comment by drphilmarshall** [comment]
@jchiang87 @rbiswas4 What's the status of this sky background study you were developing new DB capabilities for back in May?
